### PR TITLE
Added default action code for tc backend

### PR DESCRIPTION
**Pull request by Sosutha**
* Added default action code in control block for tc backend

*  Updated testcase outputs

https://github.com/Sosutha/p4c/pull/4#issuecomment-2005269828

### CODE REVIEW

**Comment by Sosutha** [comment]
@jhsmt  @vbnogueira  Please review

**Review by vbnogueira — APPROVED** [review]
LGTM

File: backends/tc/ebpfCodeGen.cpp [hunk 1]
```diff
@@ -965,16 +960,68 @@ void EBPFTablePNA::emitAction(EBPF::CodeBuilder *builder, cstring valueName,
...
         }
+        bool generateDefaultMissCode = false;
+        for (auto tbl : tcIR->tcPipeline->tableDefs) {
+            if (tbl->getTableName() == table->container->name.originalName) {
+                if (tbl->isTcMayOverride) {
+                    cstring defaultActionName = tbl->defaultMissAction->getActionName();
+                    auto actionNameStr = defaultActionName.c_str();
+                    for (long unsigned int i = 0; i < defaultActionName.size(); i++) {
+                        if (actionNameStr[i] == '/') {
+                            defaultActionName =
```
Comment (psivanup) [inline backends/tc/ebpfCodeGen.cpp]: will cstring's find/ findlast solve the purpose?

**Review by psivanup — APPROVED** [review]
Looks fine, except the good to have comment.

File: backends/tc/ebpfCodeGen.cpp (hunk 1 above)
Comment (Sosutha) [reply backends/tc/ebpfCodeGen.cpp]: Both these functions returns pointer to the character in the string, but not actual position, hence they cannot be used

**Review by maheswari-s — APPROVED** [review]
Looks fine

File: backends/tc/ebpfCodeGen.cpp (hunk 1 above)
Comment (psivanup) [reply backends/tc/ebpfCodeGen.cpp]: ref: test_backend.cpp
            fieldString = fieldString.substr(fieldString.find('.') - fieldString.begin() + 1);

Comment (Sosutha) [reply backends/tc/ebpfCodeGen.cpp]: Updated

File: testdata/p4tc_samples_outputs/tc_may_override_example_01_control_blocks.c [hunk 2]
```diff
@@ -111,9 +111,18 @@ if (/* hdr->ipv4.isValid() */
                         switch (value->action) {
                             case MAINCONTROLIMPL_IPV4_TBL_1_ACT_MAINCONTROLIMPL_NEXT_HOP: 
                                 {
+                                    if (value->is_default_miss_act) 
+                                    {
+/* send_to_port(p4tc_filter_fields.ipv4_tbl_1_next_hop_vport) */
+                                        compiler_meta__->drop = false;
```
Comment (usha1830) [inline testdata/p4tc_samples_outputs/tc_may_override_example_01_control_blocks.c:117]: this statement and the other common statements in if/else can be moved outside if

**Review by usha1830 — APPROVED** [review]
Looks fine to me. Just see if we can bring the common code in generated C files out of the if/else

File: testdata/p4tc_samples_outputs/tc_may_override_example_01_control_blocks.c (hunk 2 above)
Comment (Sosutha) [reply testdata/p4tc_samples_outputs/tc_may_override_example_01_control_blocks.c:117]: Moving the common statements out of if/else requires more common code from ebpf to be replicated. Hence implemented like this

**Pull request closed without merging**